### ISSUE TITLE
sasmodels documentation does not build without warnings.

### ISSUE THREAD

**Issue by andyfaff** [SasView/sasmodels#407]
```
cd doc
make clean
make html
```
-->
```
reading sources... [100%] model/vesicle                                                                                                                    
WARNING: error while formatting arguments for sasmodels.bumps_model.Experiment:
Traceback (most recent call last):
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/events.py", line 110, in emit
    results.append(listener.handler(self.app, *args))
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/ext/autodoc/type_comment.py", line 120, in update_annotations_using_type_comments
    type_sig = get_type_comment(obj, bound_method)
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/ext/autodoc/type_comment.py", line 108, in get_type_comment
    return signature_from_ast(subject, bound_method, function)  # type: ignore
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/ext/autodoc/type_comment.py", line 79, in signature_from_ast
    params[i] = param.replace(annotation=type_comment.argtypes[i])  # type: ignore
IndexError: list index out of range

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/ext/autodoc/__init__.py", line 419, in format_signature
    args = self._call_format_args(**kwargs)
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/ext/autodoc/__init__.py", line 404, in _call_format_args
    return self.format_args()
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/ext/autodoc/__init__.py", line 1382, in format_args
    self._signature_class, self._signature_method_name, sig = self._get_signature()
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/ext/autodoc/__init__.py", line 1355, in _get_signature
    self.env.app.emit('autodoc-before-process-signature', init, True)
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/application.py", line 450, in emit
    return self.events.emit(event, *args, allowed_exceptions=allowed_exceptions)
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/events.py", line 117, in emit
    raise ExtensionError(__("Handler %r for event %r threw an exception") %
sphinx.errors.ExtensionError: Handler <function update_annotations_using_type_comments at 0x7f8813fd6dc0> for event 'autodoc-before-process-signature' threw an exception
WARNING: error while formatting arguments for sasmodels.compare.compare:
Traceback (most recent call last):
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/events.py", line 110, in emit
    results.append(listener.handler(self.app, *args))
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/ext/autodoc/type_comment.py", line 120, in update_annotations_using_type_comments
    type_sig = get_type_comment(obj, bound_method)
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/ext/autodoc/type_comment.py", line 108, in get_type_comment
    return signature_from_ast(subject, bound_method, function)  # type: ignore
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/ext/autodoc/type_comment.py", line 79, in signature_from_ast
    params[i] = param.replace(annotation=type_comment.argtypes[i])  # type: ignore
IndexError: list index out of range

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/ext/autodoc/__init__.py", line 419, in format_signature
    args = self._call_format_args(**kwargs)
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/ext/autodoc/__init__.py", line 404, in _call_format_args
    return self.format_args()
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/ext/autodoc/__init__.py", line 1166, in format_args
    self.env.app.emit('autodoc-before-process-signature', self.object, False)
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/application.py", line 450, in emit
    return self.events.emit(event, *args, allowed_exceptions=allowed_exceptions)
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/events.py", line 117, in emit
    raise ExtensionError(__("Handler %r for event %r threw an exception") %
sphinx.errors.ExtensionError: Handler <function update_annotations_using_type_comments at 0x7f8813fd6dc0> for event 'autodoc-before-process-signature' threw an exception
WARNING: error while formatting arguments for sasmodels.compare.make_engine:
Traceback (most recent call last):
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/events.py", line 110, in emit
    results.append(listener.handler(self.app, *args))
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/ext/autodoc/type_comment.py", line 120, in update_annotations_using_type_comments
    type_sig = get_type_comment(obj, bound_method)
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/ext/autodoc/type_comment.py", line 108, in get_type_comment
    return signature_from_ast(subject, bound_method, function)  # type: ignore
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/ext/autodoc/type_comment.py", line 79, in signature_from_ast
    params[i] = param.replace(annotation=type_comment.argtypes[i])  # type: ignore
IndexError: list index out of range

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/ext/autodoc/__init__.py", line 419, in format_signature
    args = self._call_format_args(**kwargs)
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/ext/autodoc/__init__.py", line 404, in _call_format_args
    return self.format_args()
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/ext/autodoc/__init__.py", line 1166, in format_args
    self.env.app.emit('autodoc-before-process-signature', self.object, False)
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/application.py", line 450, in emit
    return self.events.emit(event, *args, allowed_exceptions=allowed_exceptions)
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/events.py", line 117, in emit
    raise ExtensionError(__("Handler %r for event %r threw an exception") %
sphinx.errors.ExtensionError: Handler <function update_annotations_using_type_comments at 0x7f8813fd6dc0> for event 'autodoc-before-process-signature' threw an exception
WARNING: error while formatting arguments for sasmodels.compare.plot_models:
Traceback (most recent call last):
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/events.py", line 110, in emit
    results.append(listener.handler(self.app, *args))
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/ext/autodoc/type_comment.py", line 120, in update_annotations_using_type_comments
    type_sig = get_type_comment(obj, bound_method)
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/ext/autodoc/type_comment.py", line 108, in get_type_comment
    return signature_from_ast(subject, bound_method, function)  # type: ignore
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/ext/autodoc/type_comment.py", line 79, in signature_from_ast
    params[i] = param.replace(annotation=type_comment.argtypes[i])  # type: ignore
IndexError: list index out of range

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/ext/autodoc/__init__.py", line 419, in format_signature
    args = self._call_format_args(**kwargs)
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/ext/autodoc/__init__.py", line 404, in _call_format_args
    return self.format_args()
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/ext/autodoc/__init__.py", line 1166, in format_args
    self.env.app.emit('autodoc-before-process-signature', self.object, False)
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/application.py", line 450, in emit
    return self.events.emit(event, *args, allowed_exceptions=allowed_exceptions)
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/events.py", line 117, in emit
    raise ExtensionError(__("Handler %r for event %r threw an exception") %
sphinx.errors.ExtensionError: Handler <function update_annotations_using_type_comments at 0x7f8813fd6dc0> for event 'autodoc-before-process-signature' threw an exception
WARNING: error while formatting arguments for sasmodels.compare.plot_profile:
Traceback (most recent call last):
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/events.py", line 110, in emit
    results.append(listener.handler(self.app, *args))
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/ext/autodoc/type_comment.py", line 120, in update_annotations_using_type_comments
    type_sig = get_type_comment(obj, bound_method)
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/ext/autodoc/type_comment.py", line 108, in get_type_comment
    return signature_from_ast(subject, bound_method, function)  # type: ignore
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/ext/autodoc/type_comment.py", line 79, in signature_from_ast
    params[i] = param.replace(annotation=type_comment.argtypes[i])  # type: ignore
IndexError: list index out of range

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/ext/autodoc/__init__.py", line 419, in format_signature
    args = self._call_format_args(**kwargs)
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/ext/autodoc/__init__.py", line 404, in _call_format_args
    return self.format_args()
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/ext/autodoc/__init__.py", line 1166, in format_args
    self.env.app.emit('autodoc-before-process-signature', self.object, False)
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/application.py", line 450, in emit
    return self.events.emit(event, *args, allowed_exceptions=allowed_exceptions)
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/events.py", line 117, in emit
    raise ExtensionError(__("Handler %r for event %r threw an exception") %
sphinx.errors.ExtensionError: Handler <function update_annotations_using_type_comments at 0x7f8813fd6dc0> for event 'autodoc-before-process-signature' threw an exception
WARNING: error while formatting arguments for sasmodels.compare.randomize_pars:
Traceback (most recent call last):
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/events.py", line 110, in emit
    results.append(listener.handler(self.app, *args))
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/ext/autodoc/type_comment.py", line 120, in update_annotations_using_type_comments
    type_sig = get_type_comment(obj, bound_method)
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/ext/autodoc/type_comment.py", line 108, in get_type_comment
    return signature_from_ast(subject, bound_method, function)  # type: ignore
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/ext/autodoc/type_comment.py", line 79, in signature_from_ast
    params[i] = param.replace(annotation=type_comment.argtypes[i])  # type: ignore
IndexError: list index out of range

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/ext/autodoc/__init__.py", line 419, in format_signature
    args = self._call_format_args(**kwargs)
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/ext/autodoc/__init__.py", line 404, in _call_format_args
    return self.format_args()
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/ext/autodoc/__init__.py", line 1166, in format_args
    self.env.app.emit('autodoc-before-process-signature', self.object, False)
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/application.py", line 450, in emit
    return self.events.emit(event, *args, allowed_exceptions=allowed_exceptions)
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/events.py", line 117, in emit
    raise ExtensionError(__("Handler %r for event %r threw an exception") %
sphinx.errors.ExtensionError: Handler <function update_annotations_using_type_comments at 0x7f8813fd6dc0> for event 'autodoc-before-process-signature' threw an exception
WARNING: error while formatting arguments for sasmodels.compare.run_models:
Traceback (most recent call last):
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/events.py", line 110, in emit
    results.append(listener.handler(self.app, *args))
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/ext/autodoc/type_comment.py", line 120, in update_annotations_using_type_comments
    type_sig = get_type_comment(obj, bound_method)
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/ext/autodoc/type_comment.py", line 108, in get_type_comment
    return signature_from_ast(subject, bound_method, function)  # type: ignore
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/ext/autodoc/type_comment.py", line 79, in signature_from_ast
    params[i] = param.replace(annotation=type_comment.argtypes[i])  # type: ignore
IndexError: list index out of range

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/ext/autodoc/__init__.py", line 419, in format_signature
    args = self._call_format_args(**kwargs)
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/ext/autodoc/__init__.py", line 404, in _call_format_args
    return self.format_args()
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/ext/autodoc/__init__.py", line 1166, in format_args
    self.env.app.emit('autodoc-before-process-signature', self.object, False)
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/application.py", line 450, in emit
    return self.events.emit(event, *args, allowed_exceptions=allowed_exceptions)
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/events.py", line 117, in emit
    raise ExtensionError(__("Handler %r for event %r threw an exception") %
sphinx.errors.ExtensionError: Handler <function update_annotations_using_type_comments at 0x7f8813fd6dc0> for event 'autodoc-before-process-signature' threw an exception
docstring of sasmodels.compare:87: WARNING: Unexpected indentation.
WARNING: error while formatting arguments for sasmodels.data.empty_data1D:
Traceback (most recent call last):
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/events.py", line 110, in emit
    results.append(listener.handler(self.app, *args))
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/ext/autodoc/type_comment.py", line 120, in update_annotations_using_type_comments
    type_sig = get_type_comment(obj, bound_method)
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/ext/autodoc/type_comment.py", line 108, in get_type_comment
    return signature_from_ast(subject, bound_method, function)  # type: ignore
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/ext/autodoc/type_comment.py", line 79, in signature_from_ast
    params[i] = param.replace(annotation=type_comment.argtypes[i])  # type: ignore
IndexError: list index out of range

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/ext/autodoc/__init__.py", line 419, in format_signature
    args = self._call_format_args(**kwargs)
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/ext/autodoc/__init__.py", line 404, in _call_format_args
    return self.format_args()
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/ext/autodoc/__init__.py", line 1166, in format_args
    self.env.app.emit('autodoc-before-process-signature', self.object, False)
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/application.py", line 450, in emit
    return self.events.emit(event, *args, allowed_exceptions=allowed_exceptions)
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/events.py", line 117, in emit
    raise ExtensionError(__("Handler %r for event %r threw an exception") %
sphinx.errors.ExtensionError: Handler <function update_annotations_using_type_comments at 0x7f8813fd6dc0> for event 'autodoc-before-process-signature' threw an exception
WARNING: error while formatting arguments for sasmodels.data.load_data:
Traceback (most recent call last):
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/events.py", line 110, in emit
    results.append(listener.handler(self.app, *args))
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/ext/autodoc/type_comment.py", line 120, in update_annotations_using_type_comments
    type_sig = get_type_comment(obj, bound_method)
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/ext/autodoc/type_comment.py", line 108, in get_type_comment
    return signature_from_ast(subject, bound_method, function)  # type: ignore
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/ext/autodoc/type_comment.py", line 79, in signature_from_ast
    params[i] = param.replace(annotation=type_comment.argtypes[i])  # type: ignore
IndexError: list index out of range

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/ext/autodoc/__init__.py", line 419, in format_signature
    args = self._call_format_args(**kwargs)
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/ext/autodoc/__init__.py", line 404, in _call_format_args
    return self.format_args()
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/ext/autodoc/__init__.py", line 1166, in format_args
    self.env.app.emit('autodoc-before-process-signature', self.object, False)
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/application.py", line 450, in emit
    return self.events.emit(event, *args, allowed_exceptions=allowed_exceptions)
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/events.py", line 117, in emit
    raise ExtensionError(__("Handler %r for event %r threw an exception") %
sphinx.errors.ExtensionError: Handler <function update_annotations_using_type_comments at 0x7f8813fd6dc0> for event 'autodoc-before-process-signature' threw an exception
WARNING: error while formatting arguments for sasmodels.details.dispersion_mesh:
Traceback (most recent call last):
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/events.py", line 110, in emit
    results.append(listener.handler(self.app, *args))
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/ext/autodoc/type_comment.py", line 120, in update_annotations_using_type_comments
    type_sig = get_type_comment(obj, bound_method)
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/ext/autodoc/type_comment.py", line 108, in get_type_comment
    return signature_from_ast(subject, bound_method, function)  # type: ignore
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/ext/autodoc/type_comment.py", line 79, in signature_from_ast
    params[i] = param.replace(annotation=type_comment.argtypes[i])  # type: ignore
IndexError: list index out of range

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/ext/autodoc/__init__.py", line 419, in format_signature
    args = self._call_format_args(**kwargs)
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/ext/autodoc/__init__.py", line 404, in _call_format_args
    return self.format_args()
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/ext/autodoc/__init__.py", line 1166, in format_args
    self.env.app.emit('autodoc-before-process-signature', self.object, False)
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/application.py", line 450, in emit
    return self.events.emit(event, *args, allowed_exceptions=allowed_exceptions)
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/events.py", line 117, in emit
    raise ExtensionError(__("Handler %r for event %r threw an exception") %
sphinx.errors.ExtensionError: Handler <function update_annotations_using_type_comments at 0x7f8813fd6dc0> for event 'autodoc-before-process-signature' threw an exception
WARNING: error while formatting arguments for sasmodels.mixture.make_mixture_info:
Traceback (most recent call last):
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/events.py", line 110, in emit
    results.append(listener.handler(self.app, *args))
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/ext/autodoc/type_comment.py", line 120, in update_annotations_using_type_comments
    type_sig = get_type_comment(obj, bound_method)
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/ext/autodoc/type_comment.py", line 108, in get_type_comment
    return signature_from_ast(subject, bound_method, function)  # type: ignore
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/ext/autodoc/type_comment.py", line 79, in signature_from_ast
    params[i] = param.replace(annotation=type_comment.argtypes[i])  # type: ignore
IndexError: list index out of range

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/ext/autodoc/__init__.py", line 419, in format_signature
    args = self._call_format_args(**kwargs)
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/ext/autodoc/__init__.py", line 404, in _call_format_args
    return self.format_args()
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/ext/autodoc/__init__.py", line 1166, in format_args
    self.env.app.emit('autodoc-before-process-signature', self.object, False)
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/application.py", line 450, in emit
    return self.events.emit(event, *args, allowed_exceptions=allowed_exceptions)
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/events.py", line 117, in emit
    raise ExtensionError(__("Handler %r for event %r threw an exception") %
sphinx.errors.ExtensionError: Handler <function update_annotations_using_type_comments at 0x7f8813fd6dc0> for event 'autodoc-before-process-signature' threw an exception
WARNING: error while formatting arguments for sasmodels.sesans.SesansTransform:
Traceback (most recent call last):
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/events.py", line 110, in emit
    results.append(listener.handler(self.app, *args))
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/ext/autodoc/type_comment.py", line 120, in update_annotations_using_type_comments
    type_sig = get_type_comment(obj, bound_method)
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/ext/autodoc/type_comment.py", line 108, in get_type_comment
    return signature_from_ast(subject, bound_method, function)  # type: ignore
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/ext/autodoc/type_comment.py", line 79, in signature_from_ast
    params[i] = param.replace(annotation=type_comment.argtypes[i])  # type: ignore
IndexError: list index out of range

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/ext/autodoc/__init__.py", line 419, in format_signature
    args = self._call_format_args(**kwargs)
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/ext/autodoc/__init__.py", line 404, in _call_format_args
    return self.format_args()
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/ext/autodoc/__init__.py", line 1382, in format_args
    self._signature_class, self._signature_method_name, sig = self._get_signature()
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/ext/autodoc/__init__.py", line 1355, in _get_signature
    self.env.app.emit('autodoc-before-process-signature', init, True)
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/application.py", line 450, in emit
    return self.events.emit(event, *args, allowed_exceptions=allowed_exceptions)
  File "/Users/andrew/miniconda3/envs/dev3/lib/python3.8/site-packages/sphinx/events.py", line 117, in emit
    raise ExtensionError(__("Handler %r for event %r threw an exception") %
sphinx.errors.ExtensionError: Handler <function update_annotations_using_type_comments at 0x7f8813fd6dc0> for event 'autodoc-before-process-signature' threw an exception
looking for now-outdated files... none found
pickling environment... done
checking consistency... done
preparing documents... done
writing output... [100%] model/vesicle                                                                                                                     
/Users/andrew/Documents/Andy/programming/sasmodels/doc/guide/plugin.rst:13: WARNING: undefined label: add/multiply models (if the link has no caption the label must precede a section header)
generating indices...  genindex py-modindexdone
writing additional pages...  searchdone
copying images... [100%] model/img/vesicle_autogenfig.png                                                                                                  
copying downloadable files... [100%] model/src/vesicle.c                                                                                                   
copying static files... ... done
copying extra files... done
dumping search index in English (code: en)... done
dumping object inventory... done
build succeeded, 14 warnings.
```

**Comment by pkienzle** [comment]
Works for me and works on [travis](https://travis-ci.org/github/SasView/sasmodels/jobs/697275711).

Maybe a sphinx version issue?
```
$ python -m sphinx --version
__main__.py 2.4.0
```

**Comment by andyfaff** [comment]
```
andrew$ python -m sphinx --version
__main__.py 3.1.0
```

**Comment by llimeht** [comment]
With sphinx 2.4.3 the following warnings were generate:

> WARNING: error while formatting arguments for sasmodels.bumps_model.Experiment: list index out of range
> WARNING: error while formatting arguments for sasmodels.compare.compare: list index out of range
> WARNING: error while formatting arguments for sasmodels.compare.make_engine: list index out of range
> WARNING: error while formatting arguments for sasmodels.compare.plot_models: list index out of range
> WARNING: error while formatting arguments for sasmodels.compare.plot_profile: list index out of range
> WARNING: error while formatting arguments for sasmodels.compare.randomize_pars: list index out of range
> WARNING: error while formatting arguments for sasmodels.compare.run_models: list index out of range
> docstring of sasmodels.compare:87: WARNING: Unexpected indentation.
> WARNING: error while formatting arguments for sasmodels.data.empty_data1D: list index out of range
> WARNING: error while formatting arguments for sasmodels.data.load_data: list index out of range
> WARNING: error while formatting arguments for sasmodels.details.dispersion_mesh: list index out of range
> WARNING: error while formatting arguments for sasmodels.mixture.make_mixture_info: list index out of range
> WARNING: error while formatting arguments for sasmodels.sesans.SesansTransform: list index out of range

These now seem to be generating errors instead of warnings with newer sphinx.

We talked a while about whether we should ensure that sphinx warnings were treated as errors in the context of sasview documentation. I guess once these errors are fixed, it would be handy to give sphinx `-W --keep-going`.

**Comment by pkienzle** [comment]
I'm not seeing the errors, but I'm guessing it is because type signatures haven't been maintained.

For example, in [sasmodels.bumps_model.Experiment](https://github.com/SasView/sasmodels/blob/fa39f3dd2a0e8530aa61269f6abd83ab31f4d435/sasmodels/bumps_model.py#L141-L142) try changing the type signature to:
```python
from typing import ..., Optional
...

def __init__(self, data, model, cutoff=1e-5, name=None, extra_pars=None):
    # type: (Data, Model, float, Optional[str], Optional[Dict[str, Parameter]]) -> None
```

Running the mypy tests as part of the build would help identify these.

**Comment by pkienzle** [comment]
Please try PR #408 and see if it cleans up the doc errors.

Getting sasmodels to pass mypy will be a lot of work.  It's currently showing 500+ errors. Judging by the few issues that I tried to resolve there are going to be some benefits (the types of the variables will be clearer), but there's also going to be a lot of spurious code to make the type checker happy.

Numpy master branch just recently got type stubs added to the repo, so we may want to wait for the next release before type checking.

**Comment by llimeht** [comment]
PR #408 stops sphinx 3.1 crashing when generating the documentation, so is a definite improvement.

There are two remaining errors in the .rst that cause problems:
 * `sasmodels/compare.py:119` is invalid rst due to the indenting of the list of module types. The rst error can be fixed by just removing the indent; however, that entire docstring needs to be rethought as it is being used as both plain text cli output and also rst; when used as rst, it currently generates gibberish HTML as seen in the [SasView docs](http://www.sasview.org/docs/dev/sasmodels-api/sasmodels.html#module-sasmodels.compare)
 * `doc/guide/plugin.rst:13` is an invalid cross-reference; it is referencing a section in the SasView manual but that is not available when sasmodels is being built. It might be possible to create a stub Add/Multiple Models section within the sasmodels documentation that satisfies the xref and then omit that stub if the documentation is being built within sasview.

If we can come up with workable solutions to those two, then here's a patch that [turns sphinx warnings into errors](https://github.com/llimeht/sasmodels/commit/fd0602eeab8f32a7ea5d5879f01da82a466ed280) which would then let [travis-ci help catch problems](https://travis-ci.org/github/llimeht/sasmodels/jobs/697491656). (I've not figured out why travis is not warning about the type annotations currently; I suspect additional autodoc extensions might need to be installed.)

**Comment by pkienzle** [comment]
I "solved" the `compare.py` issue by not including the usage docs in the manual. Not ideal, but the other option is too much work: converting to python's command line parsing flavour of the day and hoping it can generate rst, man and cli suitable docs.

I "solved" the sasview back ref by removing the link.  Also not ideal.

My version of sphinx was giving warnings whenever I used the `# type: (...) -> return` syntax, so sadly I changed it to the less comprehensible single line syntax rather than tagging each parameter individually.

I turned on "nitpick" mode (the `-n` option to the sphinx command) and found a bunch more cross-reference errors.  Not sure why it wasn't picking them up before.

I've turned sphinx warnings into errors as requested. I've got mixed feeling about this—I'm concerned that sphinx is going to be too fragile, but I suppose it is better than having the errors silently ignored.
